File: app/api/ontology.py
"""
本体工作台控制器，提供本体对象和数据抽取任务相关API接口
"""

import json
import logging
import asyncio
from fastapi import APIRouter, Query, Request
from fastapi.responses import StreamingResponse

from app.services.ontology.dto import (
    OntologyObjectUpdate, OntologyObjectBatchCreate, OntologyTaskCreate, OntologyTaskUpdate
)
from app.services.ontology.service import OntologyService
from app.core.ontology import OntologyObjectCore, OntologyTaskCore
from app.utils.response import ResponseUtil, ApiResponse
from app.constants.ontology_constants import (
    OntologyTaskStatus,
    ONTOLOGY_TASK_STREAM_PREFIX, ONTOLOGY_TASK_STATUS_PREFIX,
    ONTOLOGY_TASK_REDIS_EXPIRE
)
from app.database.redis_utils import redis_utils
from app.database.models import OntologyTask

router = APIRouter()
logger = logging.getLogger(__name__)


# ==================== 本体对象 API ====================

@router.get("/object/list", response_model=ApiResponse)
def get_ontology_objects(
    datasource_id: str = Query(None, description="数据源ID"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量"),
    sort_by: str = Query('name', description="排序字段：name/title/description/created_at"),
    sort_order: str = Query('asc', description="排序方式：asc/desc"),
    name: str = Query(None, description="名称模糊查询")
):
    """获取本体对象列表"""
    objects, total = OntologyService.get_ontology_objects(datasource_id, page, page_size, sort_by, sort_order, name)
    return ResponseUtil.success(data={"data": objects, "total": total})


@router.get("/object/{object_id}", response_model=ApiResponse)
def get_ontology_object(object_id: str):
    """获取单个本体对象"""
    obj = OntologyService.get_ontology_object(object_id)
    if not obj:
        return ResponseUtil.not_found(message="本体对象不存在")
    return ResponseUtil.success(data=obj)


@router.post("/object/batch", response_model=ApiResponse)
def batch_create_ontology_objects(dto: OntologyObjectBatchCreate):
    """批量创建本体对象"""
    results = OntologyService.batch_create_ontology_objects(dto)
    return ResponseUtil.success(
        data=results,
        message=f"成功创建{len(results)}个本体对象"
    )


@router.post("/object/batch_delete", response_model=ApiResponse)
def batch_delete_ontology_objects(data: dict):
    """批量删除本体对象"""
    object_ids = data.get('object_ids', [])
    if not object_ids:
        return ResponseUtil.error(code=400, message="请选择要删除的本体对象")
    deleted_count = OntologyService.batch_delete_ontology_objects(object_ids)
    return ResponseUtil.success(message=f"成功删除{deleted_count}个本体对象")


@router.post("/object/batch_export", response_model=ApiResponse)
def batch_export_ontology_objects(data: dict):
    """批量导出本体对象元数据到一个文件"""
    object_ids = data.get('object_ids', [])
    export_format = data.get('export_format', 'json')
    if not object_ids:
        return ResponseUtil.error(code=400, message="请选择要导出的本体对象")
    content = OntologyObjectCore.batch_export_ontology_metadata(object_ids, export_format)
    if content is None:
        return ResponseUtil.error(code=400, message="导出失败")
    return ResponseUtil.success(data={"content": content, "format": export_format})


@router.post("/object/{object_id}", response_model=ApiResponse)
def update_ontology_object(object_id: str, dto: OntologyObjectUpdate):
    """更新本体对象"""
    obj = OntologyService.update_ontology_object(object_id, dto)
    if not obj:
        return ResponseUtil.not_found(message="本体对象不存在")
    return ResponseUtil.success(data=obj, message="更新成功")


@router.post("/object/{object_id}/delete", response_model=ApiResponse)
def delete_ontology_object(object_id: str):
    """删除本体对象"""
    success = OntologyService.delete_ontology_object(object_id)
    if not success:
        return ResponseUtil.not_found(message="本体对象不存在")
    return ResponseUtil.success(message="删除成功")


@router.post("/object/{object_id}/sync", response_model=ApiResponse)
def sync_ontology_object(object_id: str):
    """同步本体对象字段（调用core层）"""
    obj = OntologyObjectCore.sync_ontology_object(object_id)
    if not obj:
        return ResponseUtil.not_found(message="本体对象不存在")
    return ResponseUtil.success(data=obj, message="同步成功")


@router.get("/object/{object_id}/query", response_model=ApiResponse)
def query_ontology_data(
    object_id: str,
    limit: int = Query(10, ge=1, le=100, description="查询条数"),
    custom_sql: str = Query(None, description="自定义SQL语句")
):
    """查询本体对象数据（调用core层，支持自定义SQL）"""
    result = OntologyObjectCore.query_ontology_data(object_id, limit, custom_sql)
    if result.get('success'):
        return ResponseUtil.success(data=result.get('data'), message=result.get('message', '查询成功'))
    return ResponseUtil.error(code=400, message=result.get('message', '查询失败'))


@router.get("/object/{object_id}/export", response_model=ApiResponse)
def export_ontology_metadata(object_id: str, export_format: str = Query(..., description="导出格式：json/markdown")):
    """导出本体对象元数据（调用core层）"""
    content = OntologyObjectCore.export_ontology_metadata(object_id, export_format)
    if content is None:
        return ResponseUtil.not_found(message="本体对象不存在")
    return ResponseUtil.success(data={"content": content, "format": export_format})


# ==================== 导出格式 ====================

@router.get("/export_formats", response_model=ApiResponse)
def get_export_formats():
    """获取导出格式列表"""
    return ResponseUtil.success(data=OntologyService.get_export_formats())


# ==================== 数据抽取任务 API ====================

@router.get("/task/list", response_model=ApiResponse)
def get_tasks(
    name: str = Query(None, description="任务名称（模糊查询）"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量")
):
    """获取任务列表"""
    tasks, total = OntologyService.get_tasks(name, page, page_size)
    return ResponseUtil.success(data={"data": tasks, "total": total})


@router.get("/task/{task_id}", response_model=ApiResponse)
def get_task(task_id: str):
    """获取单个任务"""
    task = OntologyService.get_task(task_id)
    if not task:
        return ResponseUtil.not_found(message="任务不存在")
    return ResponseUtil.success(data=task)


@router.post("/task", response_model=ApiResponse)
def create_task(dto: OntologyTaskCreate):
    """创建数据抽取任务"""
    task = OntologyService.create_task(dto)
    return ResponseUtil.success(data=task, message="任务创建成功")


@router.post("/task/{task_id}", response_model=ApiResponse)
def update_task(task_id: str, dto: OntologyTaskUpdate):
    """更新数据抽取任务"""
    task = OntologyService.update_task(task_id, dto)
    if not task:
        return ResponseUtil.error(code=400, message="任务不存在或当前状态不可编辑")
    return ResponseUtil.success(data=task, message="任务更新成功")


@router.post("/task/batch_delete", response_model=ApiResponse)
def batch_delete_tasks(data: dict):
    """批量删除任务"""
    task_ids = data.get('task_ids', [])
    if not task_ids:
        return ResponseUtil.bad_request(message="请选择要删除的任务")
    deleted_count = OntologyService.batch_delete_tasks(task_ids)
    return ResponseUtil.success(message=f"成功删除{deleted_count}个任务")


@router.post("/task/{task_id}/start")
async def start_task(task_id: str):
    """启动任务（流式接口，调用core层执行）"""
    task = OntologyTask.select().where(
        OntologyTask.id == task_id,
        OntologyTask.deleted == False
    ).first()
    if not task:
        return ResponseUtil.not_found(message="任务不存在")

    if task.status == OntologyTaskStatus.RUNNING:
        return ResponseUtil.bad_request(message="任务正在运行中")

    success = OntologyTaskCore.execute_task(task_id)
    if not success:
        return ResponseUtil.error(code=400, message="启动任务失败")

    return ResponseUtil.success(message="任务已启动")


@router.post("/task/{task_id}/stop")
def stop_task(task_id: str):
    """停止任务（调用core层）"""
    success = OntologyTaskCore.stop_task(task_id)
    if not success:
        return ResponseUtil.bad_request(message="任务未在运行中")

    return ResponseUtil.success(message="任务已停止")


@router.post("/task/{task_id}/rerun")
async def rerun_task(task_id: str):
    """重新执行任务"""
    return await start_task(task_id)


@router.get("/task/{task_id}/stream")
async def stream_task(task_id: str):
    """SSE流式获取任务状态与进度（调用core层获取Redis键）"""
    stream_key, status_key = OntologyTaskCore.get_stream_keys(task_id)

    async def event_generator():
        try:
            if redis_utils.is_available:
                buffered = redis_utils.client.lrange(stream_key, 0, -1) or []
                last_len = len(buffered)
                for raw in buffered:
                    if isinstance(raw, bytes):
                        raw = raw.decode('utf-8')
                    yield f"data: {raw}\n\n"
                    await asyncio.sleep(0)

                while True:
                    status = redis_utils.get(status_key) or ''
                    current_len = redis_utils.client.llen(stream_key) or 0

                    if current_len > last_len:
                        new_raw = redis_utils.client.lrange(stream_key, last_len, -1) or []
                        for raw in new_raw:
                            if isinstance(raw, bytes):
                                raw = raw.decode('utf-8')
                            yield f"data: {raw}\n\n"
                            await asyncio.sleep(0)
                        last_len = current_len

                    if status in (OntologyTaskStatus.DONE, OntologyTaskStatus.FAIL, OntologyTaskStatus.CANCEL):
                        if current_len > last_len:
                            continue
                        yield "data: [DONE]\n\n"
                        return

                    await asyncio.sleep(0.5)

            else:
                yield f"data: {json.dumps({'type': 'error', 'message': 'Redis不可用'})}\n\n"
                yield "data: [DONE]\n\n"

        except asyncio.CancelledError:
            raise
        except Exception as e:
            logger.error(f"SSE流式异常: task_id={task_id}, error={e}")
            yield "data: [DONE]\n\n"

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        }
    )


@router.get("/task/{task_id}/result", response_model=ApiResponse)
def get_task_result(task_id: str):
    """获取任务执行结果（调用core层）"""
    result = OntologyTaskCore.get_task_result(task_id)
    if result is None:
        return ResponseUtil.not_found(message="任务不存在")
    return ResponseUtil.success(data=result)


@router.delete("/task/{task_id}", response_model=ApiResponse)
def delete_task(task_id: str):
    """删除单个任务"""
    task = OntologyTask.select().where(
        OntologyTask.id == task_id,
        OntologyTask.deleted == False
    ).first()
    if not task:
        return ResponseUtil.not_found(message="任务不存在")
    if task.status == OntologyTaskStatus.RUNNING:
        return ResponseUtil.bad_request(message="任务正在运行中，无法删除")
    task.delete_instance()
    return ResponseUtil.success(message="删除成功")

File: web/src/services/ontology.ts
/**
 * 本体工作台服务
 * 提供本体对象和数据抽取任务相关的API调用
 */

import http from '../utils/request';
import { API_BASE_URL } from '../utils/request';

export interface OntologyForeignKey {
  referenced_table: string;
  referenced_column: string;
}

export interface OntologyColumn {
  column_name: string;
  column_name_cn: string;
  column_description: string;
  data_type: string;
  is_primary_key: boolean;
  is_nullable: boolean;
  foreign_key: OntologyForeignKey | null;
}

export interface OntologyContent {
  table_name: string;
  title?: string;
  description?: string;
  columns: OntologyColumn[];
}

export interface OntologyObject {
  id: string;
  datasource_id: string;
  name: string;
  title: string;
  description: string;
  content: OntologyContent;
  created_at: string;
  updated_at: string;
}

export interface OntologyObjectBatchItem {
  name: string;
  title?: string;
  description?: string;
  content?: OntologyContent;
}

export interface OntologyTask {
  id: string;
  name: string;
  datasource_id: string;
  datasource_name?: string;
  configs: {
    ontology_object_id?: string;
    custom_sql?: string;
    export_format?: string;
    [key: string]: any;
  };
  status: string;
  status_label: string;
  task_progress: number;
  task_progress_message: string;
  task_begin_at: string;
  task_end_at: string;
  task_duration: number;
  created_at: string;
  updated_at: string;
}

export interface ExportFormat {
  value: string;
  label: string;
  sample: string;
}

export interface TaskResult {
  status: string;
  status_label: string;
  has_result: boolean;
  file_name?: string;
  format?: string;
  file_base64?: string;
  row_count?: number;
  executed_at?: string;
  expire_at?: string;
  message?: string;
  task_progress?: number;
  task_progress_message?: string;
  task_begin_at?: string;
  task_end_at?: string;
  task_duration?: number;
}

export const ontologyService = {
  // ==================== 本体对象 ====================

  /** 获取本体对象列表 */
  getObjects: async (datasourceId?: string, page: number = 1, pageSize: number = 20, sortBy: string = 'name', sortOrder: string = 'asc', name?: string): Promise<{ data: OntologyObject[]; total: number }> => {
    let params = [`page=${page}`, `page_size=${pageSize}`, `sort_by=${sortBy}`, `sort_order=${sortOrder}`];
    if (datasourceId) params.push(`datasource_id=${datasourceId}`);
    if (name) params.push(`name=${encodeURIComponent(name)}`);
    return http.get(`/aicenter/v1/ontology/object/list?${params.join('&')}`);
  },

  /** 获取单个本体对象 */
  getObject: async (id: string): Promise<OntologyObject> => {
    return http.get(`/aicenter/v1/ontology/object/${id}`);
  },

  /** 批量创建本体对象 */
  batchCreateObjects: async (datasourceId: string, objects: OntologyObjectBatchItem[]): Promise<OntologyObject[]> => {
    return http.post('/aicenter/v1/ontology/object/batch', { datasource_id: datasourceId, objects });
  },

  /** 更新本体对象 */
  updateObject: async (id: string, data: { title?: string; description?: string; content?: OntologyContent }): Promise<OntologyObject> => {
    return http.post(`/aicenter/v1/ontology/object/${id}`, data);
  },

  /** 删除本体对象 */
  deleteObject: async (id: string): Promise<void> => {
    return http.post(`/aicenter/v1/ontology/object/${id}/delete`);
  },

  /** 批量删除本体对象 */
  batchDeleteObjects: async (objectIds: string[]): Promise<void> => {
    return http.post('/aicenter/v1/ontology/object/batch_delete', { object_ids: objectIds });
  },

  /** 批量导出本体对象元数据 */
  batchExportObjects: async (objectIds: string[], exportFormat: string): Promise<{ content: string; format: string }> => {
    return http.post('/aicenter/v1/ontology/object/batch_export', { object_ids: objectIds, export_format: exportFormat });
  },

  /** 同步本体对象字段 */
  syncObject: async (id: string): Promise<OntologyObject> => {
    return http.post(`/aicenter/v1/ontology/object/${id}/sync`);
  },

  /** 查询本体对象数据 */
  queryObjectData: async (id: string, limit: number = 10, customSql?: string): Promise<{ columns: string[]; rows: any[]; total: number }> => {
    let params = [`limit=${limit}`];
    if (customSql) params.push(`custom_sql=${encodeURIComponent(customSql)}`);
    return http.get(`/aicenter/v1/ontology/object/${id}/query?${params.join('&')}`);
  },

  /** 导出本体对象元数据 */
  exportObjectMetadata: async (id: string, format: string): Promise<{ content: string; format: string }> => {
    return http.get(`/aicenter/v1/ontology/object/${id}/export?export_format=${format}`);
  },

  // ==================== 导出格式 ====================

  /** 获取导出格式列表 */
  getExportFormats: async (): Promise<{ formats: ExportFormat[] }> => {
    return http.get('/aicenter/v1/ontology/export_formats');
  },

  // ==================== 数据抽取任务 ====================

  /** 获取任务列表 */
  getTasks: async (name?: string, page: number = 1, pageSize: number = 20, datasourceId?: string, taskStatus?: string): Promise<{ data: OntologyTask[]; total: number }> => {
    let params = [`page=${page}`, `page_size=${pageSize}`];
    if (name) params.push(`name=${encodeURIComponent(name)}`);
    if (datasourceId) params.push(`datasource_id=${datasourceId}`);
    if (taskStatus) params.push(`task_status=${taskStatus}`);
    return http.get(`/aicenter/v1/ontology/task/list?${params.join('&')}`);
  },

  /** 获取单个任务 */
  getTask: async (id: string): Promise<OntologyTask> => {
    return http.get(`/aicenter/v1/ontology/task/${id}`);
  },

  /** 创建任务 */
  createTask: async (data: {
    name: string;
    datasource_id: string;
    configs: Record<string, any>;
  }): Promise<OntologyTask> => {
    return http.post('/aicenter/v1/ontology/task', data);
  },

  /** 更新任务 */
  updateTask: async (id: string, data: {
    name?: string;
    datasource_id?: string;
    configs?: Record<string, any>;
  }): Promise<OntologyTask> => {
    return http.post(`/aicenter/v1/ontology/task/${id}`, data);
  },

  /** SSE任务事件流地址（状态实时推送） */
  getTaskEventsUrl: (): string => {
    return `${API_BASE_URL}/aicenter/v1/ontology/task/events`;
  },

  /** SSE流式获取任务进度 */
  getTaskStreamUrl: (id: string): string => {
    return `${API_BASE_URL}/aicenter/v1/ontology/task/${id}/stream`;
  },

  /** 启动任务 */
  startTask: async (id: string): Promise<void> => {
    return http.post(`/aicenter/v1/ontology/task/${id}/start`);
  },

  /** 重新执行任务 */
  rerunTask: async (id: string): Promise<void> => {
    return http.post(`/aicenter/v1/ontology/task/${id}/rerun`);
  },

  /** 停止任务 */
  stopTask: async (id: string): Promise<void> => {
    return http.post(`/aicenter/v1/ontology/task/${id}/stop`);
  },

  /** 获取任务结果 */
  getTaskResult: async (id: string): Promise<TaskResult> => {
    return http.get(`/aicenter/v1/ontology/task/${id}/result`);
  },

  /** 下载本体任务结果文件（通过任务中心统一下载接口，支持本体任务ID或任务中心任务ID） */
  downloadTaskResult: async (taskId: string): Promise<{ blob: Blob; fileName: string }> => {
    const response = await fetch(`${API_BASE_URL}/aicenter/v1/task_center/task/${taskId}/download`);
    if (!response.ok) {
      let errMsg = `下载失败 (HTTP ${response.status})`;
      try {
        const body = await response.json();
        if (body?.message) errMsg = body.message;
      } catch {}
      throw new Error(errMsg);
    }
    const blob = await response.blob();
    // 从Content-Disposition解析后端文件名（含扩展名），作为下载文件名
    let fileName = '';
    const disposition = response.headers.get('Content-Disposition') || '';
    const match = disposition.match(/filename="?([^";]+)"?/i);
    if (match?.[1]) {
      try { fileName = decodeURIComponent(match[1]); } catch { fileName = match[1]; }
    }
    return { blob, fileName };
  },

  /** 获取本体任务结果文件下载URL（直链下载，浏览器不弹出保存对话框） */
  getTaskDownloadUrl: (ontologyTaskId: string): string => {
    return `${API_BASE_URL}/aicenter/v1/task_center/task/${ontologyTaskId}/download`;
  },

  /** 批量删除任务 */
  batchDeleteTasks: async (taskIds: string[]): Promise<void> => {
    return http.post('/aicenter/v1/ontology/task/batch_delete', { task_ids: taskIds });
  },

  /** 批量执行任务（跳过运行中/等待执行的任务，加入队列） */
  batchExecuteTasks: async (taskIds: string[]): Promise<{ success: string[]; skipped: { task_id: string; message: string }[]; failed: { task_id: string; message: string }[] }> => {
    return http.post('/aicenter/v1/ontology/task/batch_execute', { task_ids: taskIds });
  },

  /** 删除单个任务 */
  deleteTask: async (id: string): Promise<void> => {
    return http.del(`/aicenter/v1/ontology/task/${id}`);
  },
};